pkg/controller/aws/cache/managed.go

```diff
@@ -0,0 +1,249 @@
+	"github.com/crossplaneio/crossplane/pkg/meta"
+	"github.com/crossplaneio/crossplane/pkg/resource"
+	"github.com/crossplaneio/crossplane/pkg/util"
+)
+
+// Error strings.
+const (
+	errNewClient                = "cannot create new ElastiCache client"
+	errNotReplicationGroup      = "managed resource is not an ElastiCache replication group"
+	errDescribeReplicationGroup = "cannot describe ElastiCache replication group"
+	errGenerateAuthToken        = "cannot generate ElastiCache auth token"
+	errCreateReplicationGroup   = "cannot create ElastiCache replication group"
+	errModifyReplicationGroup   = "cannot modify ElastiCache replication group"
+	errDescribeCacheCluster     = "cannot describe ElastiCache cache cluster"
+	errDeleteReplicationGroup   = "cannot delete ElastiCache replication group"
+)
+
+// Note this is the length of the generated random byte slice before base64
+// encoding, which adds ~33% overhead. ElastiCache allows auth tokens to be
+const maxAuthTokenData = 32
+
+// AddManaged adds a controller that reconciles ReplicationGroup managed
+// resources claims by creating ElastiCache replication groups in AWS.
+func AddManaged(mgr manager.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1alpha1.ReplicationGroupGroupVersionKind),
+		resource.WithExternalConnecter(&connecter{client: mgr.GetClient()}))
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1alpha1.ReplicationGroupKind, v1alpha1.Group))
+	c, err := controller.New(name, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return errors.Wrap(err, "cannot create Kubernetes controller")
+	}
+
+	return c.Watch(&source.Kind{Type: &v1alpha1.ReplicationGroup{}}, &handler.EnqueueRequestForObject{})
+}
+
+type connecter struct {
+	client client.Client
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (resource.ExternalClient, error) {
+	g, ok := mg.(*v1alpha1.ReplicationGroup)
+	if !ok {
+		return nil, errors.New(errNotReplicationGroup)
+	}
+
+	p := &awsv1alpha1.Provider{}
+	n := meta.NamespacedNameOf(g.Spec.ProviderReference)
+	if err := c.client.Get(ctx, n, p); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider %s", n)
+	}
+
+	s := &corev1.Secret{}
+	n = types.NamespacedName{Namespace: p.GetNamespace(), Name: p.Spec.Secret.Name}
+	if err := c.client.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider secret %s", n)
+	}
+
+	client, err := elasticache.NewClient(s.Data[p.Spec.Secret.Key], p.Spec.Region)
+	return &external{client: client}, errors.Wrap(err, errNewClient)
+}
+
+type external struct{ client elasticache.Client }
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) {
+	g, ok := mg.(*v1alpha1.ReplicationGroup)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotReplicationGroup)
+	}
+
+	dr := e.client.DescribeReplicationGroupsRequest(elasticache.NewDescribeReplicationGroupsInput(g))
```

Comment:
Thinking about our idempotent `ExternalClient` made me realise there's a case in which:

1. An external resource named `coolresource` already exists. Let's say it's managed by Terraform.
1. A Crossplane user creates a managed resource and specifies that its external name should be exactly `coolresource`.
1. We try to reconcile the managed resource. `Observe` reports that it already exists.
1. We've now effectively adopted the external resource, and will run `Update` against it, potentially fighting with Terraform.

Our methods for naming external resources are currently inconsistent. Some controllers use the UID of the managed resource with a prefix, roughly `cloudsql-some-uuid-goes-here`. These controllers should be relatively immune to this issue. Other controllers support a "name format" per https://github.com/crossplaneio/crossplane/issues/458, which allows the managed resource author to explicitly configure the name of an external resource.

I'm not sure we need to take an action on this in this PR, but we may wish to raise an issue to track dealing with this kind of conflict.

Reply:
Well, this flow coincides almost 100% with out bring-your-own-device use case. For such cases, we'd recommend reclaim policy as `Retain` but nothing specific to updates exists. 

I am not sure if we want crossplane should have an ability to operate on the resources in a way that accepts them immutable in all ways. It could get to a point where we use it, but neither update nor delete it, however, then crossplane or apps that run on crossplane can be affected by other entities that work on the same resource, like Terraform. We might want to consider leaving the responsibility of that to the user.

Reply:
I have opened https://github.com/crossplaneio/crossplane/issues/628 to advance the discussion on bring your own resource topic. I think we got good amount of overlap between resource naming concerns over existing resources and https://github.com/crossplaneio/crossplane/issues/628